src/mds/Server.cc

```diff
@@ -7929,6 +7929,23 @@ void Server::handle_client_unlink(MDRequestRef& mdr)
   if (!dn)
     return;
 
+  CDentry::linkage_t *dnl = dn->get_linkage(client, mdr);
+  ceph_assert(!dnl->is_null());
+  CInode *in = dnl->get_inode();
+
+  // wait the previous unlink finshes thoroughly.
+  CDentry* _dn = in->get_projected_parent_dn();
```

Comment:
Keep in mind that the parent dn may not be projected at all.

Reply:
> This will try to get the primary dentry only, which is possibly still projected after the previous unlink request.

Understood, but my point is that if the CInode is not projected, then I do not believe the journal can be relied on to trigger reintegration?

>From my reading from the code only when the primary dentry is unlinked should it trigger reintegrate ?

I'm not sure to be honest. That's one reason I'm concerned about waiting on a queue.

Reply:
> > This will try to get the primary dentry only, which is possibly still projected after the previous unlink request.
> 
> Understood, but my point is that if the CInode is not projected, then I do not believe the journal can be relied on to trigger reintegration?

As you know the `get_projected_parent_dn()` will always return a `primary` dentry even not projected.

> 
> > From my reading from the code only when the primary dentry is unlinked should it trigger reintegrate ?
> 
> I'm not sure to be honest. That's one reason I'm concerned about waiting on a queue.

If my understanding is correct, your concern should be the previous `unlink's` projected parent `stray` dentry has possibly been popped from projected list and later a new `link` request comes and projects it, or there have multiple `projected parent` dentries in the projected list and the newest one is not the one we want ?

This won't happen because the `link` request will wait for the `linkmerge` to finish, please see https://github.com/ceph/ceph/blob/main/src/mds/Server.cc#L7385-L7396.

And for the `rename` request in case the `srcdn` is a remote dn for the same `CInode` it won't touch the `parent` or the `projected parent` for the `CInode`. While for the case the `dstdn` also the same. Because before the `linkmerge` to finish the `parent` or `projected parent` must be a `stray` dentry.

Did I miss something here ?